### PR TITLE
CAMEL-13071: camel-elasticsearch-rest - Add scroll api support

### PR DESCRIPTION
**Pull request by ludovic-boutros**
Hi all,

On some projects we need to fetch all results from Elasticsearch.

This is currently quite difficult.

Here is an example of usage that could be implemented:
```java
from("direct:search")
  .to("elasticsearch-rest://elasticsearch?operation=Search&indexName=twitter&indexType=tweet&useScroll=true&scrollKeepAliveMs=30000")
  .split()
  .body()
  .streaming()
  .to("mock:output")
  .end();
```

Two new parameters could be used:
 * useScroll: enable scroll usage
 * scrollKeepAliveMs: time in ms during which elasticsearch will keep search context alive
here is a proposal to implement Elasticsearch scroll API support.

### CODE REVIEW

File: components/camel-elasticsearch-rest/src/main/java/org/apache/camel/component/elasticsearch/ElasticsearchConfiguration.java [hunk 1]
```diff
@@ -51,6 +51,10 @@
     private boolean disconnect;
     @UriParam(defaultValue = "false")
     private boolean enableSSL;
+    @UriParam(defaultValue = "false")
+    private boolean useScroll;
+    @UriParam(defaultValue = "" + ElasticsearchConstants.DEFAULT_SCROLL_KEEP_ALIVE_MS)
+    private int scrollKeepAliveMs;
```
Comment (oscerd) [inline components/camel-elasticsearch-rest/src/main/java/org/apache/camel/component/elasticsearch/ElasticsearchConfiguration.java]: If you declare the default scroll keep alive as default Value you have to set on the variable too.

Comment (ludovic-boutros) [reply components/camel-elasticsearch-rest/src/main/java/org/apache/camel/component/elasticsearch/ElasticsearchConfiguration.java]: Ok, I just pushed a fix for this and code style.
Thank you @oscerd.

**Comment by oscerd** [comment]
ah, I didn't notice you're working on camel-2.23.x, can you please create the PR against master? Then we can still add the functionality to the next camel release in 2.x, but we'll backport after merging on master. Just take care of the refactoring we did for camel 3 initial work. Sorry, I really didn't notice it. It should be straightforward to do the same stuff on master..

**Comment by oscerd** [comment]
If you can't or don't have the time, just tell me and I will try to move your contributions to master.

**Comment by ludovic-boutros** [comment]
@oscerd, now this PR is against master. I fixed some code style issue as well (basically import order).
There is a test which does not pass in master (with or without this patch): [testSearchWithStringQuery](https://github.com/apache/camel/blob/e0cd8927bd57ebedbc7cf2b6c2e1e37bcc9f46aa/components/camel-elasticsearch-rest/src/test/java/org/apache/camel/component/elasticsearch/ElasticsearchGetSearchDeleteExistsUpdateTest.java#L102).

```shell
[WARNING] Flakes:
[WARNING] org.apache.camel.component.elasticsearch.ElasticsearchGetSearchDeleteExistsUpdateTest.testSearchWithStringQuery(org.apache.camel.component.elasticsearch.ElasticsearchGetSearchDeleteExistsUpdateTest)
[ERROR]   Run 1: ElasticsearchGetSearchDeleteExistsUpdateTest.testSearchWithStringQuery:117->Assert.assertEquals:645->Assert.failNotEquals:834->Assert.fail:88 response hits should be == 0 expected:<0> but was:<4>
```
But I think this should be fixed in another PR.
There's an ignored NPE as well during aggregation purge task:

```shell
[2019-01-16T15:39:12,923][WARN ][o.a.c.p.a.AggregateProcessor$AggregationTimeoutMap] [unknown] Exception occurred during purge task. This exception will be ignored.
java.lang.NullPointerException: null
	at org.apache.camel.processor.aggregate.AggregateProcessor$AggregationTimeoutMap.purge(AggregateProcessor.java:1120) ~[camel-core-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.apache.camel.support.DefaultTimeoutMap.run(DefaultTimeoutMap.java:160) [camel-core-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_181]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:1.8.0_181]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_181]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:1.8.0_181]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_181]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_181]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
```

But, I suppose this should be fixed in another PR as well.

Is it good for you ?

Thanks.

**Comment by ludovic-boutros** [comment]
Excellent ! Thanks a lot.

**Comment by oscerd** [comment]
Merged on Master, I'll backport to camel-2.x, so you'll have this one on 2.24.0

**Comment by oscerd** [comment]
Thanks!

**Pull request closed without merging**